src/BaseBuilder.js

```diff
@@ -485,33 +608,38 @@ export default class BaseBuilder {
     }
 
     const hasConcreteTemplateFactory = this._state.hasIn(['options', 'factory']);
+    const templateProviderCallback = this._getBuilderFunction('_templateProviderCallback');
 
     if (!hasConcreteTemplateFactory
-        && this._state.get('_templateProviderCallback')
+        && templateProviderCallback
         && !provider
         && !disableTemplates) {
       throw new Error('A template callback function was specified, but no provider was set');
     }
 
     // Recursively build up the options object from every field.
-    const fields = this._state.get('_fieldBuilders').entrySeq().reduce((acc, entry) => {
-      const key = entry[0];
-      const field = entry[1];
-
-      return acc.setIn(['options', 'fields', key],
-        field.getOptions(provider, { disableTemplates }));
-    }, Immutable.Map()).toJS();
+    const fields = this._state.get('_fieldBuilders').reduce((acc, field, key) => (
+      acc.setIn(['options', 'fields', key], field.getOptions(provider, { disableTemplates }))
+    ), Immutable.Map()).toJS();
 
     // If a template callback exists, realize the template, merge any
     // sub-fields, and return the resulting options object.
-    const templateProviderCallback = this._state.get('_templateProviderCallback');
     const options = !hasConcreteTemplateFactory && templateProviderCallback && !disableTemplates
       ? this._state.mergeDeep({ options: { factory: templateProviderCallback(provider) } })
       : this._state;
 
+    const flattenFunctions = functions => functions
+      .filter(value => !!value.get('name'))
+      .reduce((acc, value, key) => acc.set(key, value.get('value')), Immutable.Map());
+
+    const optionsFunctions = flattenFunctions(this._state.get('optionsFunctions'));
+    const configFunctions = flattenFunctions(this._state.get('configFunctions'));
+
     return options
       .mergeDeep(fields)
       .get('options')
+      .merge(optionsFunctions.isEmpty() ? {} : optionsFunctions)
```

Comment:
This conditional shouldn't be necessary, right? If they are empty, flattenFunctions will return a map, which will get converted to an obj by toJS()

Reply:
nice catches, thx!